In [6]:
#import neccessary library 
import numpy as np
import pandas as pd 
import requests 

## **Collect Data**

In [14]:
data = pd.read_csv('historical_air_quality_2021_en.csv')
data.head()

Station ID AQI index                Location  \
0      8767.0    102.00    10.782978,106.700711   
1      8688.0    221.00  21.0811211,105.8180306   
2      8641.0    281.00  21.0215063,105.8188748   
3     13012.0     36.00    13.998599,107.996482   
4     12488.0     68.00     16.46226,107.596351   

                                        Station name  \
0  Ho Chi Minh City US Consulate, Vietnam (Lãnh s...   
1  United Nations International School of Hanoi, ...   
2  Hanoi US Embassy, Vietnam (Đại sứ quán Mỹ, Hà ...   
3        Gia Lai/phường Thống Nhất - Pleiku, Vietnam   
4              Thừa Thiên Huế/83 Hùng Vương, Vietnam   

                                                 Url Dominent pollutant    CO  \
0  https://aqicn.org/city/vietnam/ho-chi-minh-cit...               pm25     -   
1          https://aqicn.org/city/vietnam/hanoi/unis                aqi     -   
2    https://aqicn.org/city/vietnam/hanoi/us-embassy               pm25  0.30   
3  https://aqicn.org/city/vietnam/gia-lai/phuong-...               pm25  3.00   
4  https://aqicn.org/city/vietnam/thua-thien-hue/...               pm25  2.00   

   Dew  Humidity   NO2  ...  Pressure   PM10   PM2.5   SO2 Temperature  Wind  \
0  NaN      83.0     -  ...  1,009.00    NaN  102.00     -        27.0   3.6   
1  NaN      77.0     -  ...  1,015.00    NaN       -     -        18.0   1.5   
2  NaN      77.0  8.70  ...  1,015.00    NaN  281.00  2.10        18.0   1.5   
3  NaN      60.2  6.00  ...    923.70  29.00   36.00  1.00        24.3   1.0   
4  NaN      88.0     -  ...  1,015.00  52.00   68.00     -        21.0   1.0   

           Data Time S Data Time Tz  Status Alert level  
0  2021-01-21 19:00:00       +07:00  #NAME?      #NAME?  
1  2021-01-21 20:00:00       +07:00  #NAME?      #NAME?  
2  2021-01-21 20:00:00       +07:00  #NAME?      #NAME?  
3  2021-01-21 20:00:00       +07:00  #NAME?      #NAME?  
4  2021-01-21 19:00:00       +07:00  #NAME?      #NAME?  

[5 rows x 21 columns]

## **Data Explore**

**The number of rows and columns**

In [18]:
rows, cols = data.shape
rows, cols

(3415, 21)

In [19]:
data.columns

Index(['Station ID', 'AQI index', 'Location', 'Station name', 'Url',
       'Dominent pollutant', 'CO', 'Dew', 'Humidity', 'NO2', 'O3', 'Pressure',
       'PM10', 'PM2.5', 'SO2', 'Temperature', 'Wind', 'Data Time S',
       'Data Time Tz', 'Status', 'Alert level'],
      dtype='object')

**This `Dataset on air quality in Vietnam in 2021` has 3415 rows and 21 columns**


|Column|Meanings
| :------ | ---
| Station ID  | Index
| AQI index  | Air quality index
| Location | Location
| Station name | Name of the station
| Url | Link to Real time AQI
| Dominent pollutant |Dominant Species
| CO | CO Concentration
| Dew | Fog Index
| Humidity | Humidity Index
| NO2 | NO2 Concentration
| O3 | O3 Concentration
| Pressure | Pressure Index
| PM10 | Particulate matter of 10 micrometers or less in diameter
| PM2.5 | Particulate matter of 2.5 micrometers or less in diameter
| SO2 | SO2 Concentration
| Temperature | Temperature Index
| Wind | Wind speed
| Data Time S | Update time
| Data Time Tz | Time zone
| Status | Alert Status
| Alert level | Alert level

In [17]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3415 entries, 0 to 3414
Data columns (total 21 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Station ID          2622 non-null   float64
 1   AQI index           2618 non-null   object 
 2   Location            2622 non-null   object 
 3   Station name        2622 non-null   object 
 4   Url                 2622 non-null   object 
 5   Dominent pollutant  2114 non-null   object 
 6   CO                  2615 non-null   object 
 7   Dew                 2263 non-null   float64
 8   Humidity            2608 non-null   float64
 9   NO2                 2615 non-null   object 
 10  O3                  2608 non-null   object 
 11  Pressure            2608 non-null   object 
 12  PM10                2583 non-null   object 
 13  PM2.5               2614 non-null   object 
 14  SO2                 2614 non-null   object 
 15  Temperature         2615 non-null   float64
 16  Wind  

## **Preprocessing**

**Convert object to numeric datatypes**

In [24]:
columns = ['AQI index','Dominent pollutant', 'CO', 'NO2', 'O3', 'Pressure',\
       'PM10', 'PM2.5', 'SO2']

In [25]:
data[columns] = data[columns].astype(str).apply(lambda x: pd.to_numeric(x.str.replace('\D','',regex=True)))

In [26]:
data.describe()

Station ID     AQI index  Dominent pollutant           CO  \
count   2622.000000   2101.000000         2008.000000  2155.000000   
mean   11079.826087   6663.874346           22.885956   961.350348   
std     3548.080424   4701.377106            5.220637  1377.657636   
min     1584.000000    100.000000           10.000000    10.000000   
25%     8767.000000   3200.000000           25.000000   200.000000   
50%    13012.000000   5800.000000           25.000000   500.000000   
75%    13019.000000   9100.000000           25.000000  1000.000000   
max    13252.000000  34000.000000           25.000000  9800.000000   

               Dew     Humidity          NO2           O3       Pressure  \
count  2263.000000  2608.000000  1992.000000  1227.000000    2608.000000   
mean     22.248210    75.790529  1350.085341  1022.290139  100350.479294   
std       4.256875    16.469530  1223.114050  1108.902307    1908.443486   
min       5.500000    13.000000   100.000000    90.000000   91690.000000   
25%      20.750000    62.400000   500.000000   200.000000  100200.000000   
50%      22.900000    78.000000  1000.000000   650.000000  100900.000000   
75%      25.000000    88.500000  1900.000000  1400.000000  101200.000000   
max      29.000000   100.000000  8000.000000  9300.000000  102450.000000   

               PM10         PM2.5           SO2  Temperature         Wind  
count   1888.000000   2184.000000   1261.000000  2615.000000  2615.000000  
mean    4540.338983   6324.061355   1467.581285    26.227036     2.922141  
std     3516.254299   4571.408248   4084.941581     4.733196     3.886933  
min       50.000000     50.000000     10.000000     0.500000     0.200000  
25%     1975.000000   2975.000000    200.000000    23.000000     2.000000  
50%     4200.000000   5400.000000    500.000000    26.000000     2.800000  
75%     6300.000000   9100.000000   1200.000000    29.500000     3.600000  
max    34000.000000  31500.000000  50000.000000    39.000000   150.000000